File: src/com/easy/MaximumDepthofBinaryTree_104.java
/**
 * 
 */
package com.easy;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class MaximumDepthofBinaryTree_104 {

    // Use Recursive
    // Time Complexity: O(n)
    // Space Complexity: O(logn)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 38.8 MB, less than 94.62%
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        // int mCountLeft = maxDepth(root.left);
        // int mCountRight = maxDepth(root.right);
        // return Math.max(mCountLeft, mCountRight) + 1;
        return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;

    }

    // Use Iterative
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 1 ms, faster than 11.59%
    // Memory Usage: 38.5 MB, less than 96.24%
    public int maxDepth2(TreeNode root) {
        int mResult = 0;
        if (root == null) {
            return mResult;
        }
        Queue<TreeNode> mQueue = new LinkedList<>();
        mQueue.add(root);
        while (!mQueue.isEmpty()) {
            int size = mQueue.size();
            for (int i = 0; i < size; i++) {
                TreeNode mNode = mQueue.poll();
                if (mNode.left != null) {
                    mQueue.add(mNode.left);
                }
                if (mNode.right != null) {
                    mQueue.add(mNode.right);
                }
            }
            mResult++;
        }

        return mResult;
    }
}

File: Cpp_src/easy/maximum_depth_of_binary_tree_104.cc
#include <iostream>
#include <queue>
using namespace std;

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// Use DFS
// Time Complexity: O(n)
// Space Complexity: O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 19.5 MB, less than 63.74%
class Solution
{
public:
    int maxDepth(TreeNode *root)
    {
        if (root == NULL)
        {
            return 0;
        }

        return max(maxDepth(root->left), maxDepth(root->right)) + 1;
    }
};

// Use BFS
// Time Complexity: O(n)
// Space Complexity: O(1)
// Runtime: 12 ms, faster than 58.22%
// Memory Usage: 19.2 MB, less than 92.31%
class Solution2
{
public:
    int maxDepth(TreeNode *root)
    {
        if (root == NULL)
        {
            return 0;
        }

        int result = 0;
        queue<TreeNode *> node_queue;
        node_queue.push(root);
        while (!node_queue.empty())
        {
            int size = node_queue.size();
            for (int i = 0; i < size; i++)
            {
                TreeNode *node = node_queue.front();
                node_queue.pop();
                if (node->left)
                {
                    node_queue.push(node->left);
                }
                if (node->right)
                {
                    node_queue.push(node->right);
                }
            }
            result++;
        }

        return result;
    }
};